feat: extra JSON-LD contexts and custom JSON-LD doc loaders

kdimak opened the pull request:
Signed-off-by: Dmitriy Kinoshenko <dkinoshenko@gmail.com>

#1705

Added an ability to add supplement contexts on VC/VP parsing (in particular, JSON-LD context vocabulary extension) and custom JSON-LD document loaders (to allow in-memory contexts).

kdimak commented on pkg/doc/verifiable/credential_ldp_test.go:
```diff
@@ -208,6 +211,191 @@ func TestNewCredentialFromLinkedDataProof_JSONLD_Validation(t *testing.T) {
+	r.NotNil(vcWithLdp)
+
+	// Use extra context.
+	vcWithLdp, _, err = NewCredential(vcBytes,
```
An example of using extra remote context https://trustbloc.github.io/context/vc/examples-v1.jsonld to use custom fields like credential status type `CredentialStatusList2017`  to be used in Linked Data Signature verification data and also pass strict VC validation.

kdimak commented on pkg/doc/verifiable/credential_ldp_test.go:
```diff
@@ -208,6 +211,191 @@ func TestNewCredentialFromLinkedDataProof_JSONLD_Validation(t *testing.T) {
+	r.NotNil(vcWithLdp)
+
+	// Use extra in-memory context.
+	reader, err := ld.DocumentFromReader(strings.NewReader(`
```
The most interesting part - use in-memory context.

kdimak commented on pkg/doc/signature/jsonld/processor.go:
```diff
@@ -26,23 +26,31 @@ var logger = log.New("aries-framework/json-ld-processor")
 }
 
-// WithRemoveInvalidRDF option for removing RDFs containing any one of the given matches
-func WithRemoveInvalidRDF(matches ...string) CanonicalizationOpts {
```
`WithRemoveInvalidRDF()` is dropped. `WithRemoveAllInvalidRDF()` is used by default. If any element like `CredentialStatusList2017` should be included into canonical document, it should be explicitly supported by extra JSON-LD context (with the help of `WithExtraContext()` and optionally `WithDocumentLoader()`).

kdimak commented on pkg/doc/verifiable/credential_ldp_test.go:
```diff
@@ -208,6 +211,191 @@ func TestNewCredentialFromLinkedDataProof_JSONLD_Validation(t *testing.T) {
+	vcWithLdp, _, err = NewCredential(vcBytes,
+		WithEmbeddedSignatureSuites(sigSuite),
+		WithPublicKeyFetcher(SingleKey(pubKey, kms.ED25519)),
+		WithExtraJSONLDContexts("http://localhost:8652/dummy.jsonld"),
```
I think we can add another option for in-memory context, e.g. `WithExtraInMemoryJSONLDContext()`, where we will pass just in-memory context (and won't pass this dummy context URL like `http://localhost:8652/dummy.jsonld`). Under the hood, it will generate this URL (with e.g. UUID identifier).

troyronda replied:
right - this is similar to the embedded/inline context https://www.w3.org/TR/json-ld11/#example-7-in-line-context-definition

kdimak replied:
Filed https://github.com/hyperledger/aries-framework-go/issues/1758

sudeshrshetty commented on pkg/doc/signature/verifier/verifier.go:
```diff
@@ -83,19 +67,19 @@ func New(resolver keyResolver, suites ...SignatureSuite) (*DocumentVerifier, err
 	}
 
-	return dv.verifyObject(jsonLdObject)
+	return dv.verifyObject(jsonLdObject, append(opts, jsonld.WithRemoveAllInvalidRDF()))
```
May be I am missing something here, shouldn't we keeping invalid RDFs as it is so that verifier should fail when it encounters one.

-> signer should remove invalid RDFs before signing
-> verifier should verify as it is.

troyronda replied:
Yes - that is what are trying to do.

kdimak replied:
I will re-check it in a few hours, come back with the answer. An important question!

kdimak replied:
I changed that, now `jsonld.WithRemoveAllInvalidRDF()` is applied on demand. To facilitate VC, I added `WithJSONLDOnlyValidRDF()` option.

troyronda commented on pkg/doc/signature/jsonld/processor.go:
```diff
@@ -26,23 +26,31 @@ var logger = log.New("aries-framework/json-ld-processor")
+}
+
+// WithExtraContext option is for using extra context when doing JSON-LD operations.
+func WithExtraContext(context ...string) ProcessorOpts {
```
I guess this is sometimes called external context: https://w3c.github.io/json-ld-syntax/#interpreting-json-as-json-ld

kdimak replied:
I think this is rather a hack solution: we artificially append extra contexts before doing any JSON-LD operation like compaction or normalization in order to use enlarged JSON-LD vocabularies and be capable of recognizing extra types like `name`, `description` (see https://github.com/w3c-ccg/vc-examples/issues/22), `CredentialStatusList2017`, ...
This approach should be used if we receive JSON-LD to parse and it contains out-of-vocabulary terms we still want to support (e.g. because they were used to build a signature verification data).

kdimak replied:
Agree! It would be more correct.

troyronda commented on pkg/doc/verifiable/credential_ldp_test.go:
```diff
@@ -208,6 +211,191 @@ func TestNewCredentialFromLinkedDataProof_JSONLD_Validation(t *testing.T) {
+	vcWithLdp, _, err = NewCredential(vcBytes,
+		WithEmbeddedSignatureSuites(sigSuite),
+		WithPublicKeyFetcher(SingleKey(pubKey, kms.ED25519)),
+		WithExtraJSONLDContexts("https://trustbloc.github.io/context/vc/examples-v1.jsonld"),
```
will the tests break if this file goes away?

kdimak replied:
Yes, the same applies for `https://www.w3.org/2018/credentials/examples/v1` and `https://trustbloc.github.io/context/vc/examples-v1.jsonld` used in the tests.

troyronda replied:
yup - we should probably have a todo/followup to cache contexts locally to remove this dependency.

kdimak replied:
https://github.com/hyperledger/aries-framework-go/issues/1753

troyronda commented on pkg/doc/signature/jsonld/processor.go:
```diff
@@ -26,23 +26,31 @@ var logger = log.New("aries-framework/json-ld-processor")
+// WithExternalContext option is for definition of external context when doing JSON-LD operations.
+func WithExternalContext(context ...string) ProcessorOpts {
+	return func(opts *normalizeOpts) {
+		opts.extraContexts = context
```
for consistency, would be good to replace-all on opts.extraContexts

kdimak replied:
ah, forgot to rename that.

kdimak commented:
Last update:
- "JSON-LD Only Valid RDF" on demand, added corresponding VC option.
- used `external context` term.
- added `ProofPurpose` field to DID Doc's `Proof` struct, in was missing and caused failing linked data proof checks.

One thing to note: in case of `JsonWebSignature2020`, when we verify the signature we should use `WithJSONLDOnlyValidRDF()` option. That's because this type of proof is not defined in security context (`https://w3id.org/security/v2`) that is used to build canonical documents for proofs. And we should drop it.

kdimak commented on pkg/doc/signature/jsonld/processor.go:
```diff
@@ -97,16 +113,48 @@ func (p *Processor) GetCanonicalDocument(doc map[string]interface{}, opts ...Can
 }
 
+// AppendExtraContexts appends extra context(s) to the JSON_LD context which can have one or several contexts already.
+func AppendExtraContexts(context interface{}, extraContexts ...string) []interface{} {
```
sure, thanks!

troyronda commented:
> And we should drop it.

No - include additional context.

kdimak commented:
> > And we should drop it.
> 
> No - include additional context.

The side (e.g. https://github.com/transmute-industries/lds-jws2020) which prepared VC with `JsonWebSignature2020` LDP dropped this field when building proof's canonical document. And we should do the same - otherwise we will get "invalid signature" error.

troyronda commented:
why do you get "invalid signature" error when you include the extra context?

troyronda commented:
I guess we should temporarily make an exception for `JsonWebSignature2020` when the context is missing. Do we still have a whitelist option? @kdimak

kdimak commented:
> I guess we should temporarily make an exception for `JsonWebSignature2020` when the context is missing during validation. Do we still have a whitelist option? @kdimak

i've dropped it :)

sudeshrshetty commented:
May need @rolsonquadras 's approval for this

kdimak commented:
To make forced cleanup of invalid RDF-s in proofs case only (for both signing and verification). This way we will clean JsonWebSignature2020 not declared in https://w3id.org/security/v2
For the VC content itself - we make this cleanup on demand only  (as it was before)
this way we don’t need to apply WithJSONLDOnlyValidRDF()  option for verification of VCs with JsonWebSignature2020 LDP.
